app/src/main/java/com/vci/vectorcamapp/core/data/upload/image/ImageUploadWorker.kt

```diff
@@ -2,85 +2,491 @@ package com.vci.vectorcamapp.core.data.upload.image
+        notificationSessionTitle = "Session from $sessionDateStr"
+
+        val specimensToUpload = sessionWithSpecimens.specimens.filter {
+            it.imageUploadStatus != UploadStatus.COMPLETED
+        }
+
+        if (specimensToUpload.isEmpty()) {
+            Log.d("ImageUploadWorker", "No images to upload for session $sessionId.")
+            return WorkerResult.success()
+        }
 
-        setForeground(createForegroundInfo())
+        setForeground(showInitialSessionNotification(specimensToUpload.size))
 
-        val sessionId = inputData.getString("session_id") ?: return Result.retry()
+        var successfulUploads = 0
+        var encounteredPermanentFailure = false
 
-        try {
-            for (i in 1..TOTAL_IMAGES) {
-                // Simulate image upload
-                delay(1000)
-                Log.d("UploadWorker", "Uploading image $sessionId")
-                updateNotification(i)
+        specimensToUpload.forEachIndexed { index, specimen ->
+            notificationTotalImages = specimensToUpload.size
+            notificationCurrentImageIndex = index + 1
+
+            val imageId = 50 // TODO: UPDATE TO REAL IMAGEID
+
+            when (val result = uploadSingleSpecimen(specimen, sessionId, imageId)) {
+                is DomainResult.Success -> {
+                    updateProgressNotification("Verifying...")
+                    successfulUploads++
+            }
+                is DomainResult.Error -> {
+                    if (result.error == NetworkError.TUS_PERMANENT_ERROR) {
+                        encounteredPermanentFailure = true
+                    }
+                }
             }
+        }
+
+        showFinalStatusNotification(successfulUploads, specimensToUpload.size)
+        return if (successfulUploads == specimensToUpload.size) {
+            WorkerResult.success()
+        } else if (encounteredPermanentFailure) {
+            WorkerResult.failure()
+        } else {
+            WorkerResult.retry()
+        }
+    }
+
+    private suspend fun uploadSingleSpecimen(
+        specimen: Specimen,
+        sessionId: UUID,
+        imageId: Int
+    ): DomainResult<String, NetworkError> {
+        val (file, contentType, md5) = try {
+            val (tempFile, type) = prepareFile(context.contentResolver, specimen.imageUri, specimen.id)
+            Triple(tempFile, type, calculateMD5(tempFile))
         } catch (e: Exception) {
-            Log.d("UploadWorker", "Error during upload: ${e.message}")
-            return Result.retry()
+            if (isStopped) {
+                throw CancellationException("Worker was stopped during file preparation.", e)
+            }
+            Log.e("ImageUploadWorker", "Failed to prepare file or calculate MD5 for specimen ${specimen.id}.", e)
+            specimenRepository.updateSpecimen(specimen.copy(imageUploadStatus = UploadStatus.FAILED), sessionId)
```

Comment:
should you delete the file here? I think you should have a finally block to delete the file. Also wrap within: if (file.exists()) file.delete()